### PR TITLE
BUG: DataFrame.first_valid_index() fails if there is no valid entry.

### LINKED ISSUE #17400: df.first_valid_index() fails if there is no valid entry.
https://github.com/pandas-dev/pandas/blob/3a7f956c30528736beaae5784f509a76d892e229/pandas/core/frame.py#L3770-L3777

    import pandas as pd
    import numpy as np

    df = pd.DataFrame({'a': [np.nan,np.nan]})
    df.first_valid_index()

This code results in the following error message, which I believe is called because the boolean series returned by self.count(1) > 0 is all False.

    Traceback (most recent call last):
      File "<pyshell#10>", line 1, in <module>
        df.first_valid_index()
      File "C:\Python27\lib\site-packages\pandas\core\frame.py", line 3859, in first_valid_index
        return self.index[self.count(1) > 0][0]
      File "C:\Python27\lib\site-packages\pandas\indexes\base.py", line 1423, in __getitem__
        return getitem(key)
    IndexError: index 0 is out of bounds for axis 0 with size 0

Should there be a check where None is returned if self.count(1) > 0 is all False?
As an aside, this issue also occurs in df.last_valid_index() for the same reason.

### PR DESCRIPTION
**Pull request by skwbc**
Just checking the number of valid entries and if it's zero, return `None`.

Also fixed same issue on `DataFrame.last_valid_index()`.
Add docstrings to both methods of `DataFrame` and `Series`.

- [x] closes #17400
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v0.21.0.txt [hunk 1]
```diff
@@ -399,6 +399,7 @@ Indexing
 - Bug in ``CategoricalIndex`` reindexing in which specified indices containing duplicates were not being respected (:issue:`17323`)
 - Bug in intersection of ``RangeIndex`` with negative step (:issue:`17296`)
 - Bug in ``IntervalIndex`` where performing a scalar lookup fails for included right endpoints of non-overlapping monotonic decreasing indexes (:issue:`16417`, :issue:`17271`)
+- Bug in ``DataFrame.first_valid_index`` and ``DataFrame.last_valid_index`` when no valid entry (:issue:`17400`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: you can do

``:meth:`DataFrame.first_valid_index` `` and so on

File: pandas/core/frame.py [hunk 2]
```diff
@@ -4069,23 +4069,97 @@ def update(self, other, join='left', overwrite=True, filter_func=None,
...
 
+    def _get_valid_indices(self):
+        is_valid = self.count(1) > 0
+        return self.index[is_valid]
+
     def first_valid_index(self):
         """
-        Return label for first non-NA/null value
+        Return index for first non-NA/null value.
+        If all elements are non-NA/null, returns None.
```
Comment (jreback) [inline pandas/core/frame.py]: Add these 2 last sentences to a ``Notes`` section

Comment (jreback) [reply pandas/core/frame.py]: add a ``Returns`` section (it returns a scalar)

File: pandas/core/frame.py [hunk 3]
```diff
@@ -4069,23 +4069,97 @@ def update(self, other, join='left', overwrite=True, filter_func=None,
...
+        --------
+
+        When no null value in a DataFrame, returns first index.
+
+        >>> df = DataFrame({'A': [1, 2, 3], 'B': ['a', 'b', 'f']})
+        >>> df.first_valid_index()
+        0
+
+        When all elements in first row are null, returns second index.
+
```
Comment (jreback) [inline pandas/core/frame.py]: ideally would like to move these doc-strings to ``core/generic.py`` and use ``_shared_docs``. can you do that.

File: pandas/tests/frame/test_timeseries.py [hunk 4]
```diff
@@ -440,6 +440,11 @@ def test_first_last_valid(self):
         assert empty.last_valid_index() is None
         assert empty.first_valid_index() is None
 
+        # GH17400
```
Comment (jreback) [inline pandas/tests/frame/test_timeseries.py]: add a comment (no valid entries)

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm. see if you can consolidate the docs a bit, so we don't repeat so much.

File: pandas/core/frame.py (hunk 3 above)
Comment (skwbc) [reply pandas/core/frame.py]: @jreback 
Thank you for your advice.
Modified the doc-strings with `_shared_docs`.
Not sure to the manner of pandas doc-strings, please check it.
If it's ok, I will squash the 2nd commit to 1st one.

Note: I have no idea to share the doc-strings between `Series` and `DataFrames` with Examples, so I removed the examples.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks!

**Pull request merged**